### PR TITLE
clang-tidy, part 4

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added even more clang-tidy checks.

### CODE REVIEW

File: dbms/src/Functions/FunctionsStringRegex.cpp [hunk 1]
```diff
@@ -306,8 +306,8 @@ struct MultiMatchAnyImpl
         MultiRegexps::ScratchPtr smart_scratch(scratch);
 
         auto on_match = []([[maybe_unused]] unsigned int id,
-                           unsigned long long /* from */,
-                           unsigned long long /* to */,
+                           unsigned long long /* from */, // NOLINT
+                           unsigned long long /* to */, // NOLINT
```
Comment (qoega) [inline dbms/src/Functions/FunctionsStringRegex.cpp]: Why we do so?

File: dbms/src/Dictionaries/HashedDictionary.cpp [hunk 2]
```diff
@@ -446,7 +446,7 @@ void HashedDictionary::addAttributeSize(const Attribute & attribute)
         /** TODO: more accurate calculation */
         bytes_allocated += sizeof(CollectionType<T>);
         bytes_allocated += bucket_count;
-        bytes_allocated += map_ref->size() * sizeof(Key) * sizeof(T);
+        bytes_allocated += map_ref->size() * (sizeof(Key) + sizeof(T));
```
Comment (qoega) [inline dbms/src/Dictionaries/HashedDictionary.cpp]: There was a real bug? What should work better now?

File: dbms/src/Common/UTF8Helpers.cpp [hunk 3]
```diff
@@ -75,15 +75,10 @@ static int wcwidth(wchar_t wc)
     switch (width)
     {
         case widechar_nonprint:
-            [[fallthrough]];
         case widechar_combining:
-            [[fallthrough]];
         case widechar_unassigned:
-            return 0;
```
Comment (qoega) [inline dbms/src/Common/UTF8Helpers.cpp]: What is the logic here? Default `width` is 0?

File: dbms/src/Functions/FunctionsStringRegex.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsStringRegex.cpp]: It's the interface of Hyperscan.

And `long long` is not the synonim of `uint64_t`.

uint64_t can correspond to either long and long long.
long and long long have the same behaviour on Linux amd64 and Mac OS amd64 and FreeBSD amd64 but not on Windows amd64.

File: dbms/src/Functions/bitCount.cpp [hunk 4]
```diff
@@ -26,7 +26,7 @@ struct BitCountImpl
         if constexpr (std::is_same_v<A, Int8>)
             return __builtin_popcount(static_cast<UInt8>(a));
         else
-            return __builtin_popcountll(ext::bit_cast<unsigned long long>(a));
+            return __builtin_popcountll(ext::bit_cast<uint64_t>(a));
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply dbms/src/Functions/bitCount.cpp]: There is the same issue. 
From the name of `__builtin_popcountll` it's clear that it corresponds to the long long type.
But in this case we know that actually we need 64bit unsigned integer.

And it does not matter if we use uint64_t or UInt64.

File: dbms/src/Dictionaries/HashedDictionary.cpp (hunk 2 above)
Comment (alexey-milovidov) [reply dbms/src/Dictionaries/HashedDictionary.cpp]: Yes, looks like the amount of allocated space was not calculated correcly in system.dictionaries for sparse_hashed dictionary layout. We may ask @azat to make sure.

File: dbms/src/Common/UTF8Helpers.cpp (hunk 3 above)
Comment (alexey-milovidov) [reply dbms/src/Common/UTF8Helpers.cpp]: It's a bug, that I introduced and it was found by our test!

File: dbms/src/Storages/MergeTree/MergeTreeData.h [hunk 5]
```diff
@@ -601,7 +601,6 @@ class MergeTreeData : public IStorage
 
     /// Check that the part is not broken and calculate the checksums for it if they are not present.
     MutableDataPartPtr loadPartAndFixMetadata(const DiskPtr & disk, const String & relative_path);
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply dbms/src/Storages/MergeTree/MergeTreeData.h]: It is using another non-static method. At least clang-tidy did not suggest me to do that.

File: dbms/src/Dictionaries/HashedDictionary.cpp (hunk 2 above)
Comment (azat) [reply dbms/src/Dictionaries/HashedDictionary.cpp]: Yep, not it looks sane

FWIW `SparseCollectionType` should be used over `CollectionType` here (fortunately this does not affects the result a lot)

P.S. does `clang-tidy` automatically finds this? Interesting!

Comment (alexey-milovidov) [reply dbms/src/Dictionaries/HashedDictionary.cpp]: > Yep, not it looks sane

You mean "now it looks sane".

> FWIW SparseCollectionType should be used over CollectionType here

Ok, will fix.

> does clang-tidy automatically finds this? Interesting!

Yes, it is https://clang.llvm.org/extra/clang-tidy/checks/bugprone-sizeof-expression.html

Comment (azat) [reply dbms/src/Dictionaries/HashedDictionary.cpp]: >You mean "now it looks sane".

Yep!

**Comment by alexey-milovidov** [comment]
#6044

**Pull request merged**